ethereum/trinity
# Aggregate signature validation

### What was wrong?

Fixes #243.

### How was it fixed?

Updated the logic for validating an aggregate signature on an attestation.

Note: there is some refactoring here to expose various bits of logic to avoid duplication either in code or computation.

Summary of changes:

- Added helper `get_crosslink_committee_for_attestation`
    - Returns the given crosslink committee for the given attestation
    - Raises ValidationError if there is no such shard
    - Covered by existing test for `get_attestation_participants`
- Added helper `get_members_from_bitfield`
    - Returns the val_index from the committee if they are marked in the bitfield
    - Should not return anyone not marked in the bitfield
    - Covered by existing test for `get_attestation_participants`
- Added a helper `generate_aggregate_pubkeys_from_indices`
    - Returns the aggregate pubkey for a set of validators indicated by index
    - Tested by existing tests for `generate_aggregate_pubkeys`
- Updates `validate_attestation_aggregate_signature`
    - Ensure custody_bitfield is empty
    - Ensures aggregation_bitfield is not empty
    - Ensures that is a bit is not set in aggregate_bitfield then it is also not set in the custody_bitfield
    - Uses bls.verify_multiple to verify both sets of participants, partitioned by custody bit
- Ensure we can only process up to MAX_ATTESTATIONS in a given block
- added tests for the new bits of functionality listed above not covered by existing tests


#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](http://1.bp.blogspot.com/-8dahhaWGtcA/ULTXEhSq7mI/AAAAAAAAIXc/zBdlVxfqLs4/s1600/Walrus+picture.jpg)
